test/e2e-test/appcontext_test.go

```diff
@@ -0,0 +1,297 @@
+	)
+	var ns = corev1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: namespace}}
+
+	workload1 := cw(
+		cwWithName(cwName1),
+		cwWithContainers([]v1alpha2.Container{
+			{
+				Name:    containerName,
+				Image:   containerImage,
+				Command: []string{"sleep", "30s"},
+			},
+		}),
+	)
+
+	rawWorkload1 := runtime.RawExtension{Object: workload1}
+	co1 := comp(
+		compWithName(compName1),
+		compWithNamespace(namespace),
+		compWithWorkload(rawWorkload1),
+	)
+
+	ac1 := ac(
+		acWithName(acName1),
+		acWithNamspace(namespace),
+		acWithComps([]v1alpha2.ApplicationConfigurationComponent{
+			{
+				ComponentName: compName1,
+				Traits:        []v1alpha2.ComponentTrait{},
+			},
+		}),
+	)
+	workload2 := cw(
+		cwWithName(cwName2),
+		cwWithContainers([]v1alpha2.Container{
+			{
+				Name:    containerName,
+				Image:   containerImage,
+				Command: []string{"sleep", "30s"},
+			},
+		}),
+	)
+	rawWorkload2 := runtime.RawExtension{Object: workload2}
+	co2 := comp(
+		compWithName(compName2),
+		compWithNamespace(namespace),
+		compWithWorkload(rawWorkload2),
+	)
+	dummyApp := &v1alpha2.Application{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "dummy",
+			Namespace: namespace,
+		},
+		Spec: v1alpha2.ApplicationSpec{
+			Components: []v1alpha2.ApplicationComponent{},
+		},
+	}
+	ar1 := &v1alpha2.ApplicationRevision{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      arName1,
+			Namespace: namespace,
+		},
+		Spec: v1alpha2.ApplicationRevisionSpec{
+			Components: []*v1alpha2.Component{
+				co1,
+			},
+			ApplicationConfiguration: *ac1,
+			Application:              *dummyApp,
+		}}
+	trait2 := &v1alpha2.ManualScalerTrait{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       v1alpha2.ManualScalerTraitKind,
+			APIVersion: v1alpha2.SchemeGroupVersion.String(),
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      traitName2,
+			Namespace: namespace,
+		},
+		Spec: v1alpha2.ManualScalerTraitSpec{
+			ReplicaCount: 3,
+		},
+	}
+	rawTrait2 := runtime.RawExtension{Object: trait2}
+	ac2 := ac(
+		acWithName(acName2),
+		acWithNamspace(namespace),
+		acWithComps([]v1alpha2.ApplicationConfigurationComponent{
+			{
+				ComponentName: compName2,
+				Traits: []v1alpha2.ComponentTrait{
+					v1alpha2.ComponentTrait{
+						Trait: rawTrait2,
+					},
+				},
+			},
+		}),
+	)
+	ar2 := &v1alpha2.ApplicationRevision{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      arName2,
+			Namespace: namespace,
+		},
+		Spec: v1alpha2.ApplicationRevisionSpec{
+			ApplicationConfiguration: *ac2,
+			Components: []*v1alpha2.Component{
+				co2,
+			},
+			Application: *dummyApp,
+		}}
+	appContext := &v1alpha2.ApplicationContext{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      appContextName,
+			Namespace: namespace,
+		},
+		Spec: v1alpha2.ApplicationContextSpec{
+			ApplicationRevisionName: arName1,
+		},
+	}
+
+	BeforeEach(func() {
+		Expect(k8sClient.Create(ctx, &ns)).Should(SatisfyAny(BeNil(), &util.AlreadyExistMatcher{}))
```

Comment:
better to use a different for each ns, some test environment/settings can't handle ns recreate